File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    n, m = len(maps), len(maps[0])
    
    if maps[0][0] == 0 or maps[n-1][m-1] == 0:
        return -1
    
    dp = [[0] * m for _ in range(n)]
    dp[0][0] = 1
    
    dirt = [(1, 0), (-1, 0), (0, 1), (0, -1) ]    
    q = deque()
    q.append((0, 0))
    
    while q:
        x, y = q.popleft()
        
        if x==n-1 and y==m-1:
            return dp[x][y]
        
        for dx, dy in dirt:
            nx, ny = x+dx, y+dy
            if 0<=nx<n and 0<=ny<m and maps[nx][ny]==1:
                if dp[nx][ny] == 0:
                    dp[nx][ny] = dp[x][y]+1
                    q.append((nx, ny))
                    
    return -1

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    public int solution(int[][] maps) {
        int n = maps.length;
        int m = maps[0].length;
        int[][] dp = new int[n][m];
        int[][] dirt = {{1, 0}, {0, 1}, {0, -1}, {-1, 0}};
        Deque<int[]> q = new ArrayDeque<>();
        
        q.offer(new int[]{0, 0});
        dp[0][0] = 1;
        while (!q.isEmpty()) {
            int[] curr = q.poll();
            int x = curr[0], y = curr[1];
            for (int[] next : dirt) {
                int dx = x + next[0], dy = y + next[1];
                if (0 <= dx && dx < n && 0 <= dy && dy < m && maps[dx][dy] != 0 && dp[dx][dy] == 0) {
                    q.offer(new int[]{dx, dy});
                    dp[dx][dy] = dp[x][y] + 1;
                }
            }
        }
        int answer = dp[n-1][m-1];
        return answer == 0 ? -1 : answer;
    }
}